The job **`x86_64-gnu-llvm-21`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/27105956248/job/79995068616) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/79995068616) [(plain)](https://github.com/rust-lang/rust/commit/f3e127a943b23051f5b096cde81b18140a2c0d4c/checks/79995068616/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
---- [ui] tests/ui/layout/too-big-no-dce.rs stdout ----

error: `build-fail` test is required to pass check build, but check build failed
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/layout/too-big-no-dce.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=i686-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/layout/too-big-no-dce" "-A" "unused" "-W" "unused_attributes" "-A" "internal_features" "-A" "incomplete_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/i686-unknown-linux-gnu/native/rust-test-helpers" "-Clinker=x86_64-linux-gnu-gcc" "-O"
stdout: none
--- stderr -------------------------------
error[E0080]: attempt to shift left by `62_i32`, which would overflow
##[error]  --> /checkout/tests/ui/layout/too-big-no-dce.rs:12:31
   |
LL |     assert_valid_type::<[u32; 1 << 62]>();
   |                               ^^^^^^^ evaluation of `f::{constant#0}` failed here

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0080`.
------------------------------------------
```

</details>

--- a/library/core/src/mem/mod.rs
+++ b/library/core/src/mem/mod.rs
@@ -371,6 +371,13 @@ pub fn forget_unsized<T: ?Sized>(t: T) {
 #[rustc_const_stable(feature = "const_mem_size_of", since = "1.24.0")]
 #[rustc_diagnostic_item = "mem_size_of"]
 pub const fn size_of<T>() -> usize {
+    // By making this a constant, we also guarantee that the constant can be successfully evaluated
+    // in any program execution that actually executes `size_of`. Which is relevant because the
+    // constant can fail to evaluate if the type is too big. Someone might do something cursed where
+    // soundness relies on a certain type not being too big, and they check that by just invoking
+    // size_of on the type to ensure it exists, so if we fully DCE'd size_of calls that would be
+    // considered unsound... but by making this a constant, it participates in the usual "required
+    // consts" system, and we are safe.
     <T as SizedTypeProperties>::SIZE
 }
 

--- a/tests/ui/layout/too-big-no-dce.rs
+++ b/tests/ui/layout/too-big-no-dce.rs
@@ -1,0 +1,20 @@
+//! Ensure that we do not dead-code-eliminate `size_of` calls. That might hide "type too big"
+//! errors!
+//@ build-fail (test needs codegen)
+//@ compile-flags: -O
+//@ normalize-stderr: "\d{5}\d*" -> "NUMBER"
+
+//~? ERROR too big for the target
+
+#![crate_type = "lib"]
+
+const PTR_BITS_MINUS_1: usize = std::mem::size_of::<*const ()>() * 8 - 1;
+
+#[unsafe(no_mangle)] // ensure this gets monomorphized
+pub fn f() {
+    assert_valid_type::<[u32; 1 << PTR_BITS_MINUS_1]>();
+}
+
+pub fn assert_valid_type<T>() {
+    std::mem::size_of::<T>();
+}

--- a/tests/ui/layout/too-big-no-dce.stderr
+++ b/tests/ui/layout/too-big-no-dce.stderr
@@ -1,0 +1,8 @@
+error[E0080]: values of the type `[u32; NUMBER]` are too big for the target architecture
+  --> $SRC_DIR/core/src/mem/mod.rs:LL:COL
+   |
+   = note: evaluation of `<[u32; NUMBER] as std::mem::SizedTypeProperties>::SIZE` failed here
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0080`.

